### PR TITLE
Feat metric logging

### LINKED ISSUE #129: Metric Logging
### 📌 Feature Summary

Add metric logging in the auth to track failures and successes

### 💡 Motivation

We should enable basic metric logging to track failures and successes. The logger should track metrics across the kinds of errors as well. These metrics are useful to visualize and track our errors.

### 🧩 Suggested Implementation

Use a metrics collector to collect metrics as a middleware layer, and then also add a route `/metrics` to return the values
```python
from collections import defaultdict
from typing import Dict
import threading

class MetricsCollector:
    def __init__(self):
        self.lock = threading.Lock()
        self.metrics = defaultdict(int)

    def inc(self, key: str):
        with self.lock:
            self.metrics[key] += 1

    def get(self) -> Dict[str, int]:
        with self.lock:
            return dict(self.metrics)

metrics = MetricsCollector()
```

### 🧠 Estimated Complexity

Medium (Changes multiple files, some model updates)

### 🔗 Related Issues / PRs

_No response_

### 🧠 Additional Notes

_No response_

### PR DESCRIPTION
**Pull request by snigenigmatic**
#129 - Add metrics logging with thread-safe collector and /metrics endpoint
## 📌 Description

This PR implements comprehensive metrics logging for the PESUAuth API to enable monitoring and observability of authentication requests, errors, and system performance.

**What is the purpose of this PR?**
- Adds a thread-safe metrics collection system to track authentication successes, failures, and various error types
- Implements a new `/metrics` endpoint to expose collected metrics in JSON format
- Integrates metrics tracking throughout the FastAPI application lifecycle

**What problem does it solve?**
- Provides visibility into API usage patterns and authentication success rates
- Enables monitoring of error types and frequencies for better debugging and system health assessment
- Tracks CSRF token refresh operations for background task monitoring
- Facilitates performance analysis and capacity planning

**Background:**
The API previously had no metrics or monitoring capabilities, making it difficult to assess system health, debug issues, or understand usage patterns. This implementation provides comprehensive tracking without impacting performance.

> ℹ️ **Fixes / Related Issues**
> Fixes: #129

## 🧱 Type of Change

- [x] ✨ New feature – Adds functionality without breaking existing APIs
- [x] 📝 Documentation update – README, docstrings, OpenAPI tags, etc.
- [x] 🧪 Test suite change – Adds/updates unit, functional, or integration tests
- [x] 🕵️ Debug/logging enhancement – Adds or improves logging/debug support

## 🧪 How Has This Been Tested?

- [x] Unit Tests (`tests/unit/`)
- [ ] Functional Tests (`tests/functional/`)
- [x] Integration Tests (`tests/integration/`)
- [x] Manual Testing

> ⚙️ **Test Configuration:**
>
> - OS: Windows 11
> - Python: 3.13.2 via uv
> - [x] Docker build tested

**Testing Details:**
- **Unit Tests**: 10 comprehensive tests for MetricsCollector covering thread safety, concurrent access, edge cases, and special character handling
- **Integration Tests**: End-to-end FastAPI testing with mock authentication flows to verify metrics collection in real scenarios
- **Manual Testing**: Live API testing confirmed correct metrics tracking for success/failure scenarios and endpoint functionality

## ✅ Checklist

- [x] My code follows the [CONTRIBUTING.md](https://github.com/pesu-dev/auth/blob/main/.github/CONTRIBUTING.md) guidelines
- [x] I've performed a self-review of my changes
- [x] I've added/updated necessary comments and docstrings
- [x] I've updated relevant docs (README or endpoint docs)
- [x] No new warnings introduced
- [x] I've added tests to cover my changes
- [x] All tests pass locally (`scripts/run_tests.py`) - 
- [x] I've run linting and formatting (`pre-commit run --all-files`) - 
- [x] Docker image builds and runs correctly - 
- [x] Changes are backwards compatible (if applicable)
- [ ] Feature flags or `.env` vars updated (if applicable) - **Not applicable**
- [x] I've tested across multiple environments (if applicable)
- [x] Benchmarks still meet expected performance (`scripts/benchmark_auth.py`) - 

## 🛠️ Affected API Behaviour

- [x] `app/app.py` – Modified `/authenticate` route logic

**New API Endpoint:**
- **`/metrics`** - New GET endpoint that returns current application metrics in JSON format

### 🧩 Models
* [x] `app/models/response.py` – Used existing response model for metrics endpoint formatting

**New Files Added:**
- `app/metrics.py` - Core MetricsCollector implementation with thread-safe operations
- `app/docs/metrics.py` - OpenAPI documentation for the new /metrics endpoint  
- `tests/unit/test_metrics.py` - Comprehensive unit tests for MetricsCollector
- `tests/integration/test_metrics_integration.py` - Integration tests for metrics collection

### 🐳 DevOps & Config

* [ ] `Dockerfile` – No changes to build process
* [ ] `.github/workflows/*.yaml` – No CI/CD pipeline changes required
* [ ] `pyproject.toml` / `requirements.txt` – No new dependencies added
* [ ] `.pre-commit-config.yaml` – No linting or formatting changes

### 📊 Benchmarks & Analysis

* [ ] `scripts/benchmark_auth.py` – No changes to benchmark scripts
* [ ] `scripts/analyze_benchmark.py` – No changes to analysis tools
* [ ] `scripts/run_tests.py` – No changes to test runner

### 🎯 Metrics Endpoint in Action

image

*Live metrics collection showing authentication success/failure tracking*

### Metrics Endpoint Response Example
```json
{
  "status": true,
  "message": "Metrics retrieved successfully",
  "timestamp": "2025-08-28T15:30:45.123456+05:30",
  "metrics": {
    "auth_success_total": 150,
    "auth_failure_total": 12,
    "validation_error_total": 8,
    "pesu_academy_error_total": 5,
    "unhandled_exception_total": 0,
    "csrf_token_error_total": 2,
    "profile_fetch_error_total": 1,
    "profile_parse_error_total": 0,
    "csrf_token_refresh_success_total": 45,
    "csrf_token_refresh_failure_total": 1
  }
}
```

### 🔧 Testing Results Dashboard

image-2

*Comprehensive test suite covering unit, integration, and functional scenarios*

### Updated API Endpoints Table
| **Endpoint**    | **Method** | **Description**                                        |
|-----------------|------------|--------------------------------------------------------|
| `/`             | `GET`      | Serves the interactive API documentation (Swagger UI). |
| `/authenticate` | `POST`     | Authenticates a user using their PESU credentials.     |
| `/health`       | `GET`      | A health check endpoint to monitor the API's status.   |
| `/readme`       | `GET`      | Redirects to the project's official GitHub repository. |
| **`/metrics`**  | **`GET`**  | **Returns current application metrics and counters.**   |

**Metrics Tracked:**
1. `auth_success_total` - Successful authentication attempts
2. `auth_failure_total` - Failed authentication attempts  
3. `validation_error_total` - Request validation failures
4. `pesu_academy_error_total` - PESU Academy service errors
5. `unhandled_exception_total` - Unexpected application errors
6. `csrf_token_error_total` - CSRF token extraction failures
7. `profile_fetch_error_total` - Profile page fetch failures
8. `profile_parse_error_total` - Profile parsing errors
9. `csrf_token_refresh_success_total` - Successful background CSRF refreshes
10. `csrf_token_refresh_failure_total` - Failed background CSRF refreshes

### CODE REVIEW

File: app/docs/metrics.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                                    "csrf_token_refresh_success_total": 45,
+                                    "csrf_token_refresh_failure_total": 1,
+                                },
+                            },
+                        }
+                    }
+                }
+            },
+        }
+    },
```
Comment (aditeyabaral) [inline app/docs/metrics.py:40]: Create a model for this. The response model will also need an update.

File: app/docs/metrics.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+                            "description": (
+                                "All current application metrics including authentication counts and error rates"
+                            ),
+                            "value": {
+                                "status": True,
+                                "message": "Metrics retrieved successfully",
+                                "timestamp": "2025-08-28T15:30:45.123456+05:30",
+                                "metrics": {
+                                    "auth_success_total": 150,
+                                    "auth_failure_total": 12,
```
Comment (aditeyabaral) [inline app/docs/metrics.py:24]: We should also track how many auth requests are received, including a split for how many with and without profile data

File: app/__init__.py [hunk 3]
```diff
@@ -1 +1,5 @@
 """PESUAuth API - A simple API to authenticate PESU credentials using PESU Academy."""
+
+from app.metrics import metrics
+
+__all__ = ["metrics"]
```
Comment (aditeyabaral) [inline app/__init__.py]: Remove this; we should not want to initialize a global collector outside the entry point.

File: app/app.py [hunk 4]
```diff
@@ -116,6 +125,19 @@ async def validation_exception_handler(request: Request, exc: RequestValidationE
...
+    exc_type = type(exc).__name__.lower()
+    if "csrf" in exc_type:
+        metrics.inc("csrf_token_error_total")
+    elif "profilefetch" in exc_type:
+        metrics.inc("profile_fetch_error_total")
+    elif "profileparse" in exc_type:
+        metrics.inc("profile_parse_error_total")
+    elif "authentication" in exc_type:
+        metrics.inc("auth_failure_total")
+
```
Comment (aditeyabaral) [inline app/app.py:168]: There is a much cleaner solution. Look into a [middleware layer](https://fastapi.tiangolo.com/tutorial/middleware/#create-a-middleware). Here is some pseudo code to get you started:

```python
@app.middleware("http")
async def metrics_middleware(request: Request, call_next):
    metrics.inc("requests_total")
    start_time = time.time()

    try:
        response: Response = await call_next(request)
        latency = time.time() - start_time

        # Track successes vs failures
        if 200 <= response.status_code < 300:
            metrics.inc("requests_success")
        else:
            metrics.inc("requests_failed")
            metrics.inc(f"requests_failed_status_{response.status_code}")

        # Latency metrics
        metrics.inc("request_latency_sum", latency)

        # Also add route metrics: route = request.scope.get("route")

        return response

    except Exception as e:
        latency = time.time() - start_time
        metrics.inc("requests_failed")
        metrics.inc(f"requests_failed_exception_{type(e).__name__}")
        metrics.inc("request_latency_sum", latency)

        raise

```
Note, you will need to accordingly increment other metrics like how many with and without profile data by parsing the request.

**Review by aditeyabaral — CHANGES_REQUESTED** [review]

File: pyproject.toml
Comment (achyu-dev) [inline pyproject.toml:1]: Are these dependencies necessary to be added ?

File: scripts/benchmark/util.py
Comment (achyu-dev) [inline scripts/benchmark/util.py:1]: why is this being added? any idea?

**Review by achyu-dev — CHANGES_REQUESTED** [review]
Please do the requested changes from @aditeyabaral and me

File: scripts/benchmark/util.py
Comment (snigenigmatic) [reply scripts/benchmark/util.py:1]: These changes allow the benchmark script to test all endpoints—including /metrics and non-JSON ones like /readme—without crashing, by handling different response types and adding proper headers.

**Comment by aditeyabaral** [comment]
Thanks for this, @snigenigmatic — and sorry it sat so long.

The feature has shipped in **#157**, now merged to `dev` as v4.2.0. Closing this as superseded rather than stale, because the design that shipped is substantially the one you proposed here.

**What came from your work**

- the `MetricsCollector` shape from your PR and from your sketch in #129 — an in-process counter store with an `inc`/`get` pair — is what `app/metrics/collector.py` is, with typed metric families in place of free-string keys
- a `/metrics` endpoint exposing them, with its own entry in `app/docs/`
- the metric vocabulary: auth successes and failures, per-error-type counts, CSRF refresh tracking
- unit **and** integration tests for the collector, including the thread-safety and edge cases you covered

**What #157 changed, following the review here**

- a proper response model (`MetricsModel`) rather than a raw dict
- an HTTP middleware instead of per-route instrumentation
- no collector instance in `app/__init__.py` — the singleton is created in `app/app.py` beside the `PESUAcademy` client
- auth requests counted with the with/without-profile split that was asked for

**You were right about `scripts/benchmark/util.py`, and it is worth saying so explicitly.** That change was questioned in review with "why is this being added?" — your answer was that it let the benchmark script exercise all endpoints without crashing on non-JSON responses. That is exactly correct, and it is a genuine bug, not a nicety: `make_request` ended in an unconditional `response.json()`, so `--route readme` (a 308 to GitHub returning HTML) killed the sequential runner outright and was silently swallowed as a *failed request* by the parallel one — quietly skewing the numbers the script exists to measure. #157 includes that fix, and it is credited to this PR in the description.

**Where it went beyond the original scope**, in case it is interesting: the set grew from your ten counters to twenty-three, the biggest addition being the upstream calls to PESU Academy — `csrf_fetch`, `login`, `profile_fetch` — each with its own latency, outcome and status code. That turned out to be the most informative part. On staging right now, a profile fetch averages 2.0s against a login's 0.56s, which nothing previously measured. `/metrics` also serves the Prometheus exposition format by default (`?fmt=json` for the JSON view), which unblocks #130 without needing an exporter.

The metrics also earned their keep within minutes of existing: they showed an idle process making **two** CSRF fetches at startup and discarding one of the clients, because the refresh loop refreshed immediately instead of sleeping first. Fixed in the same PR. That is the kind of thing this feature was for, and it would not have been visible without it.

If you would like to pick something else up, the tracker is open — #130 (Grafana) is now unblocked and builds directly on this.
